lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/classify/stats/CubePercentile.java

```diff
@@ -0,0 +1,60 @@
+package edu.stanford.futuredata.macrobase.analysis.classify.stats;
+
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.Multiset;
+
+import java.util.*;
+
+/**
+ * Computes percentiles when given an array of metrics and the count of the
+ * number of times each occurs. Useful for computing percentiles on cubed data.
+ */
+public class CubePercentile {
+    private double[] counts;
+    private double[] metrics;
+
+    // Computed
+    private int numRawMetrics = 0;
+    private Multiset<Double> metricCounts;
+    private List<Double> sortedMetrics;
+
+    public CubePercentile(double[] counts, double[] metrics) {
+        this.counts = counts;
+        this.metrics = metrics;
+        computeCounts();
+    }
+
+    public double evaluate(double percentile) {
+        if (percentile >= 50.0) {
+            int numToPass = (int)((100.0 - percentile) / 100.0 * numRawMetrics);
+            int numPassed = 0;
+            for (int i = sortedMetrics.size() - 1; i >= 0; i--) {
+                numPassed += metricCounts.count(sortedMetrics.get(i));
```

Comment:
I benchmarked, and it seems like pre-calculating the running sum takes much more time than the amount of time that's saved in evaluate(). This is especially true given our use case, where we only call evaluate() twice.

Indeed, the vast majority of time is spent constructing the relevant lists and arrays, and not much time is spent evaluating percentiles (maybe a 20 to 1 ratio).

Also in our use case, straight comparisons is better than binary search because we're usually evaluating very high and very low percentiles anyway.